actions/v3/upload_sessions.go

```diff
@@ -43,71 +46,69 @@ type UploadSessionCreateResV3 struct {
+	req := uploadSessionCreateReqV3{}
 	if err := oyster_utils.ParseReqBody(c.Request(), &req); err != nil {
 		err = fmt.Errorf("Invalid request, unable to parse request body  %v", err)
 		c.Error(400, err)
 		return err
 	}
 
-	// Get session
-	uploadSession := &models.UploadSession{}
-	err := models.DB.Find(uploadSession, c.Param("id"))
-
-	defer oyster_utils.TimeTrack(time.Now(), "actions/upload_sessions: updating_session", analytics.NewProperties().
-		Set("id", uploadSession.ID).
-		Set("genesis_hash", uploadSession.GenesisHash).
-		Set("file_size_byes", uploadSession.FileSizeBytes).
-		Set("num_chunks", uploadSession.NumChunks).
-		Set("storage_years", uploadSession.StorageLengthInYears))
-
-	if err != nil {
-		oyster_utils.LogIfError(err, nil)
-		c.Error(400, err)
-		return err
-	}
-	if uploadSession == nil {
-		err := errors.New("Error finding sessions")
-		oyster_utils.LogIfError(err, nil)
+	if NumChunksLimit != -1 && req.NumChunks > NumChunksLimit {
+		err := errors.New("This broker has a limit of " + fmt.Sprint(NumChunksLimit) + " file chunks.")
+		fmt.Println(err)
 		c.Error(400, err)
 		return err
 	}
 
-	treasureIdxMap, err := uploadSession.GetTreasureIndexes()
+	alphaEthAddr, privKey, _ := EthWrapper.GenerateEthAddr()
+	uploadBeta = bucketRequest{}
+	hasBeta := req.BetaIP != ""
+	if hasBeta {
+		betaReq, err := json.Marshal(req)
+		if err != nil {
+			oyster_utils.LogIfError(err, nil)
+			c.Error(400, err)
+			return err
+		}
+
+		reqBetaBody := bytes.NewBuffer(betaReq)
 
-	if oyster_utils.DataMapStorageMode == oyster_utils.DataMapsInBadger {
-		dbID := []string{oyster_utils.InProgressDir, uploadSession.GenesisHash, oyster_utils.MessageDir}
+		// Should we be hardcoding the port?
+		betaURL := req.BetaIP + ":3000/api/v2/upload-sessions/beta"
+		betaRes, err := http.Post(betaURL, "application/json", reqBetaBody)
+		defer betaRes.Body.Close() // we need to close the connection
 
-		db := oyster_utils.GetOrInitUniqueBadgerDB(dbID)
-		if db == nil {
-			err := errors.New("error creating unique badger DB for messages")
+		if err != nil {
 			oyster_utils.LogIfError(err, nil)
 			c.Error(400, err)
 			return err
 		}
+		betaSessionRes := &uploadSessionCreateBetaResV3{}
 	}
 
-	// Update dMaps to have chunks async
-	go func() {
-		defer oyster_utils.TimeTrack(time.Now(), "actions/upload_sessions: async_datamap_updates", analytics.NewProperties().
-			Set("id", uploadSession.ID).
-			Set("genesis_hash", uploadSession.GenesisHash).
-			Set("file_size_byes", uploadSession.FileSizeBytes).
-			Set("num_chunks", uploadSession.NumChunks).
-			Set("storage_years", uploadSession.StorageLengthInYears))
-
-		models.ProcessAndStoreChunkData(req.Chunks, uploadSession.GenesisHash, treasureIdxMap,
-			models.DataMapsTimeToLive)
-	}()
-
-	return c.Render(202, actions_utils.Render.JSON(map[string]bool{"success": true}))
+	res := uploadSessionCreateResV3{
+		BatchSize:   BatchSize,
+		UploadAlpha: bucketRequest{},
+		UploadBeta:  uploadBeta,
+	}
```

Comment:
I may be mistaken but I think we had started leaning towards the client still uploading to the broker, and the broker itself passing the chunks on to S3.  @EdmundMai @AaronVasquez

Reply:
But if the bucket is public-read-write, then why can't client upload directly to the bucket rather than going via the brokernode?

Reply:
The client could upload directly to S3, but then the client would have to know how to interact with S3.  The brokernode is the one who decides whether it will use badger or S3 and the client shouldn't need to know.  

There are tradeoffs either way.  We can discuss this more.

Reply:
Yes, let's do that in slack for it

Reply:
We decided on having the client uploading to the broker, and then the broker uploading to s3 for a few reasons:

- Easier to implement, and we could launch the feature sooner
- Keeps the client "dumb" so it doesn't need to know which data store the broker is using
- Makes it easier to handle auth to s3 since we can store creds on the broker. The buckets will be private and only accessible by the broker and lambda

Reply:
I am planning to update it in the next PR